specs/AppLifecycle/Restart/restartApi.markdown

```diff
@@ -0,0 +1,259 @@
+    // The new assets have downloaded and installed, and now the app has to request a restart immediately
+    protected void UpdateInstallComplete()
+    {
+        // Checking if the assets have updated successfully and there are no pending updates
+        if (!CheckForUpdate())
+        {
+            AppRestartFailureReason result = AppLifecycle::RequestRestartNow("args for new instance");
+            switch (result)
+            {
+                case AppRestartFailureReason::RestartPending:
+                    // In this case the restart could not be completed because another restart is pending and we can try to restart again after a period of time
+                    RetryRestartAfterXSeconds();
+                    break;
+                case AppRestartFailureReason::Other:
+                    // In this case, a dialog box will appear informing of an error requesting restart
+                    ShowErrorMessage();
+                    break;
+            }
+        }
+    } 
+
+    // In this scenario, assume the app encounters an error during initialization.
+    // The app displays an error dialog, and after the user clicks OK on the dialog the app must restart immediately.
+    // This scenario simply displays logging for these use cases
+    protected void HandleInitializationError()
+    {
+        AppRestartFailureReason result = AppLifecycle::RequestRestartNow("args for new instance");
+        switch (result)
+        {
+            case AppRestartFailureReason::RestartPending:
+                Telemetry::WriteLine("Another restart is currently pending.");
+                break;
+            case AppRestartFailureReason::InvalidUser:
+                Telemetry::WriteLine("Current user is not signed in or not a valid user.");
+                break;
+            case AppRestartFailureReason::Other:
+                Telemetry::WriteLine("Failure restarting.");
+                break;
+        }
+    }
+}
+```
+
+# Telemetry
+
+This section explores the telemetry needs for the Restart feature in
+four dimensions – Usage, Reliability, Performance and Failure
+Investigation. The telemetry will include the metadata that are needed
+not only for measuring quality but also for helping with investigations
+of failures. Following is the event which can be used to capture
+telemetry information:
+
+Event Name: **AppRestart**
+
+| Metadata                | Description                                                                                       | Data type |
+| ----------------------- | ------------------------------------------------------------------------------------------------- | --------- |
+| AppRestartFailureReason | Reason an app restart failed. This will be collected when an app requests a restart now but fails | Enum      |
+| AppName                 | Name of the app which restarted                                                                   | String    |
```

Comment:
⚠️ AppName could contain PII. Do not recommend using as-is.

Reply:
App names have generally been considered non-PII. App paths contain PII, but not the executable name itself.

Reply:
Darn, I better rename `RafsIceCreamConeMaker.exe`.

(I see there's established [precedent for this](https://docs.microsoft.com/en-us/windows/privacy/required-windows-diagnostic-data-events-and-fields-2004#:~:text=appname%20the%20name%20of%20the%20app%20that%20has%20crashed.) so OK.)

Reply:
In practice, software developers do not ship a custom executable name for each customer. (`ContosoDeluxe-BobSmithFrom Denver.exe`, `ContosoDeluxe-JaneJohnsonFromChicago.exe`, `ContosoDeluxe-PatWilsonFromLosAngeles.exe`.) The executable name is the same for all customers: `ContosoDeluxe.exe`.

`RafsIceCreamConeMaker.exe` is presumably a piece of software that is marketed as "Check out this cool program  'Raf's Ice Cream Cone Maker', in which a cartoon character name Raf helps you design the perfect ice cream cone." It doesn't tell you who bought the program. Raf is the program name, not the customer name. (That's the theory at least.)

Reply:
One of my favorites, PowerGrep, ships `SetupPowerGREP522RafaelRivera.exe`. (The executable is generated server-side for licensing purposes.) Hex Rays (maker of IDA) also ships custom executables with uniquely identifying details (e.g. order number, user id). But I was also thinking of scenarios where I fire up Visual Studio and just hack on a project and don't necessarily want information about it going back to the mothership. All outliers in the grand scheme, perhaps.